src/DqtApi/DataStore/Crm/DataverseAdapter.UpdateTeacher.cs

```diff
@@ -0,0 +1,563 @@
+                    Category = "DMSImportTrn",
+                    Subject = message,
+                    Description = description,
+                    ScheduledEnd = _dataverseAdapter._clock.UtcNow
+                };
+
+                string GetDescription()
+                {
+                    var sb = new StringBuilder();
+                    sb.Append(message);
+                    return sb.ToString();
+                }
+            }
+
+
+            public dfeta_initialteachertraining CreateInitialTeacherTrainingEntity(UpdateTeacherReferenceLookupResult referenceData, Guid? Id)
+            {
+                Debug.Assert(referenceData.IttCountryId.HasValue);
+                Debug.Assert(referenceData.IttProviderId.HasValue);
+                Debug.Assert(referenceData.IttSubject1Id.HasValue);
+                Debug.Assert(referenceData.IttSubject2Id.HasValue);
+
+                var cohortYear = _command.InitialTeacherTraining.ProgrammeEndDate.Year.ToString();
+
+                return new dfeta_initialteachertraining()
+                {
+                    Id = Id.HasValue ? Id.Value : Guid.NewGuid(),
+                    dfeta_PersonId = new EntityReference(Contact.EntityLogicalName, TeacherId),
+                    dfeta_CountryId = new EntityReference(dfeta_country.EntityLogicalName, referenceData.IttCountryId.Value),
+                    dfeta_EstablishmentId = new EntityReference(Account.EntityLogicalName, referenceData.IttProviderId.Value),
+                    dfeta_ProgrammeStartDate = _command.InitialTeacherTraining.ProgrammeStartDate.ToDateTime(),
+                    dfeta_ProgrammeEndDate = _command.InitialTeacherTraining.ProgrammeEndDate.ToDateTime(),
+                    dfeta_ProgrammeType = _command.InitialTeacherTraining.ProgrammeType,
+                    dfeta_CohortYear = cohortYear,
+                    dfeta_Subject1Id = referenceData.IttSubject1Id.HasValue ? new EntityReference(dfeta_ittsubject.EntityLogicalName, referenceData.IttSubject1Id.Value) : null,
+                    dfeta_Subject2Id = referenceData.IttSubject2Id.HasValue ? new EntityReference(dfeta_ittsubject.EntityLogicalName, referenceData.IttSubject2Id.Value) : null,
+                    dfeta_Result = _command.InitialTeacherTraining.Result,
+                    dfeta_AgeRangeFrom = _command.InitialTeacherTraining.AgeRangeFrom,
+                    dfeta_AgeRangeTo = _command.InitialTeacherTraining.AgeRangeTo,
+                };
+            }
+
+            public dfeta_qualification CreateQualificationEntity(UpdateTeacherReferenceLookupResult referenceData, Guid? id)
+            {
+                Debug.Assert(referenceData.QualificationId.HasValue);
+                Debug.Assert(referenceData.QualificationCountryId.HasValue);
+                Debug.Assert(referenceData.QualificationSubjectId.HasValue);
+
+                if (id.HasValue)
```

Comment:
I think it might be neater to create entity once and remove the attributes that aren't required for the update case e.g.
```cs
var entity = new dfeta_qualification()
{
    // all the fields
};

if (id.HasValue)
{
    entity.Attributes.Remove(dfeta_qualification.Fields.dfeta_PersonId);
}

return entity;
```